File: frontend/scripts/check-deployed-revision.js
#!/usr/bin/env node
/**
 * Le frontend SERVI est-il celui de ce commit ?
 *
 * ── Le trou que ce garde ferme ─────────────────────────────────────────────
 * Le backend répond quelle révision il exécute (`/health`, champ `revision`) et
 * `deploy-fly` refuse quand ce n'est pas le commit attendu (CI-COVERAGE.md, F8).
 * Le frontend, lui, ne disait RIEN : le déploiement de production est fait par
 * Vercel depuis l'app GitHub, aucun job ne le commande, donc « main est vert »
 * ne pouvait pas distinguer « le site servi est construit depuis ce commit » de
 * « Vercel a échoué / est en retard / n'a pas redéployé ». Rien n'observait la
 * production du frontend pour le CODE — les sondes existantes y regardent le SEO,
 * les OG et le CORS, jamais de quel commit l'artefact est fait.
 *
 * Le build publie donc sa révision dans le HTML statique
 * (`vite-plugins/inject-build-revision.js` → meta `kojo-build-revision`), et ce
 * garde compare cette réponse à ce que `main` aurait dû déployer.
 *
 * ── Quelle révision est ATTENDUE ───────────────────────────────────────────
 * La propriété visée est : aucun changement de `frontend/**` n'est absent de la
 * production. Elle s'écrit avec deux valeurs acceptées, et non une seule, parce
 * que la politique du déployeur n'est pas la nôtre :
 *   • `--attendu` : le commit poussé (Vercel redéploie chaque push de `main`,
 *     mesuré sur 12 commits consécutifs) ;
 *   • `--sinon`   : le dernier commit qui a touché `frontend/**` — l'état que la
 *     production laisse en place si un jour Vercel ignore un push sans
 *     changement de frontend. L'accepter ne relâche rien : dans les deux cas, le
 *     contenu frontend servi est celui de `main`.
 * Accepter une valeur PLUS ANCIENNE ne relâcherait rien non plus, mais serait un
 * refus silencieux de la question posée : un déploiement en retard doit rougir.
 *
 * ── Ce que ce garde refuse, et pourquoi il le NOMME ────────────────────────
 * Trois refus, trois corrections différentes — un seul « échec » les confondrait :
 *   1. le HTML servi n'annonce AUCUNE révision : l'artefact est antérieur à ce
 *      mécanisme, ou construit sans qu'aucune source n'en donne (ni `KOJO_GIT_SHA`,
 *      ni variable système, ni dépôt git — cf. `vite-plugins/inject-build-revision.js`).
 *      Correction : redéployer le frontend sur Vercel.
 *   2. la révision servie n'est ni l'attendue ni la dernière ayant touché le
 *      frontend : la production est en retard (build échoué, en file d'attente
 *      depuis plus longtemps que la fenêtre d'attente, ou déploiement annulé).
 *   3. aucune réponse exploitable : le site ne répond pas, ou le CDN refuse
 *      (défi de sécurité de l'edge Vercel, mesuré le 20/09/2026 et nommé ici :
 *      `X-Vercel-Mitigated`). Rien ne peut alors être affirmé sur le déploiement.
 *
 * Une invocation qui ne dit pas quoi comparer ÉCHOUE (code 2) au lieu de passer :
 * `--attendu` vide ou trop court n'est pas un identifiant de commit.
 *
 * ── Coût, et le risque de faux rouge ───────────────────────────────────────
 * Un appel HTTP par tentative, séquentiels, bornés par `--tentatives` (défaut 9
 * espacés de 20 s ≈ 3 min) : c'est ce qui absorbe un déploiement Vercel en cours
 * sans jamais transformer un refus en succès. Le défi de l'edge Vercel (le seul
 * faux rouge observé sur la production, cf. `resolve-vercel-url.sh`) est nommé
 * comme tel, donc reconnu au lieu d'être confondu avec un frontend en retard.
 *
 * Usage :
 *   cd frontend && node scripts/check-deployed-revision.js \
 *     --attendu "$GITHUB_SHA" --sinon "$(git rev-list -1 HEAD -- frontend/)"
 * Optionnel :
 *   --base https://kojoforafrica.cc.cd   (défaut : SITE_ORIGIN, scripts/site-meta.js)
 *   --tentatives 9 --delai 20
 */
import { pathToFileURL } from 'node:url';
import { BUILD_REVISION_META, SITE_ORIGIN, isLoopbackUrl, metaContent } from './site-meta.js';

// Longueur minimale d'un SHA de commit : en dessous, ce n'est pas un identifiant
// de commit mais une chaîne quelconque — comparer là-dessus serait un faux vert.
export const LONGUEUR_SHA_MIN = 7;

// Valeurs qui ne sont PAS une révision : le champ vide, et ce qu'un artefact
// fabriqué sans révision pourrait annoncer.
export const REVISIONS_ABSENTES = new Set(['', 'inconnue', 'unknown', 'none', 'null', 'undefined', 'n/a']);

// Un seul en-tête, pour que la requête soit identifiable dans les journaux du CDN.
const PROBE_HEADERS = { 'user-agent': 'kojo-deployed-revision/1.0' };

/**
 * La valeur est-elle une RÉVISION ? Vide et « inconnue » ne le sont pas.
 *
 * Une seule fonction pour les deux côtés de la comparaison : sans ça, une
 * révision « inconnue » servie et la même attendue se déclareraient concordantes
 * — le faux vert exact que ce garde existe pour fermer.
 *
 * @param {unknown} valeur Valeur brute (chaîne, null, autre type).
 * @returns {string} La révision, ou '' si la valeur n'en est pas une.
 */
export function normaliser(valeur) {
  if (typeof valeur !== 'string') return '';
  const propre = valeur.trim();
  return REVISIONS_ABSENTES.has(propre.toLowerCase()) ? '' : propre;
}

/**
 * Les révisions acceptées, à partir de ce que l'appelant a fourni.
 *
 * @param {Array<string|undefined>} valeurs `--attendu`, puis `--sinon`.
 * @returns {{valides: string[], invalides: string[]}} Les invalides sont rendues
 *   telles quelles : c'est ce qui permet de les NOMMER dans le refus.
 */
export function revisionsAttendues(valeurs = []) {
  const valides = [];
  const invalides = [];
  for (const valeur of valeurs) {
    const brute = String(valeur ?? '').trim();
    if (!brute) continue;
    const révision = normaliser(brute);
    if (révision.length < LONGUEUR_SHA_MIN) invalides.push(brute);
    else if (!valides.includes(révision)) valides.push(révision);
  }
  return { valides, invalides };
}

/**
 * La révision publiée par le HTML servi, ou '' si la page n'en annonce aucune.
 *
 * @param {string} html Corps de la réponse de la page d'accueil.
 * @returns {string} La révision annoncée.
 */
export function revisionServie(html) {
  return normaliser(metaContent(String(html ?? ''), BUILD_REVISION_META));
}

/**
 * Le verdict, avec le coupable nommé dans chaque cas.
 *
 * @param {string} servie Révision annoncée par la production.
 * @param {string[]} attendues Révisions acceptées.
 * @returns {{ok: boolean, message: string}} Verdict et son explication.
 */
export function comparer(servie, attendues = []) {
  const révision = normaliser(servie);
  const acceptées = attendues.map(normaliser).filter(Boolean);
  const attendu = acceptées.length ? acceptées.map((r) => r.slice(0, 12)).join(' ou ') : '(aucune)';

  if (!révision) {
    return {
      ok: false,
      message:
        `le frontend servi n'annonce AUCUNE révision (meta « ${BUILD_REVISION_META} » absente) : ` +
        "l'artefact déployé est antérieur à ce mécanisme, ou il a été construit sans qu'aucune " +
        'source n\'en donne (KOJO_GIT_SHA, variables système, ou le dépôt git du build). ' +
        'Redéployer le frontend sur Vercel — tant que cette balise se tait, rien ne prouve ' +
        'quel code le site sert.',
    };
  }
  if (acceptées.includes(révision)) {
    return { ok: true, message: `révision servie = attendue (${révision.slice(0, 12)})` };
  }
  return {
    ok: false,
    message:
      `révision servie ≠ attendue : servi=${révision.slice(0, 12)} attendu=${attendu} — ` +
      "le frontend déployé n'est pas celui de main (build Vercel échoué, en file d'attente, " +
      'ou déploiement jamais déclenché).',
  };
}

/**
 * Lit la page d'accueil du site servi.
 *
 * @param {string} base Origine du site (`https://…`, sans slash final).
 * @param {object} [options]
 * @param {Function} [options.fetchImpl] `fetch` injectable (tests).
 * @param {number} [options.timeout] Délai maximal de la requête, en ms.
 * @returns {Promise<{html: string|null, erreur: string|null}>} Jamais d'exception.
 */
export async function interroger(base, { fetchImpl = fetch, timeout = 15000 } = {}) {
  let réponse;
  try {
    réponse = await fetchImpl(`${base}/`, {
      redirect: 'follow',
      headers: PROBE_HEADERS,
      signal: AbortSignal.timeout(timeout),
    });
  } catch (error) {
    return { html: null, erreur: `${error.name || 'Error'}: ${error.message}` };
  }

  if (!réponse.ok) {
    // Le défi de l'edge Vercel a déjà produit des rouges au hasard sur `main`
    // (403 `X-Vercel-Mitigated: challenge`, cf. resolve-vercel-url.sh) : le nommer
    // ici évite qu'il se lise comme « frontend en retard ».
    const mitigation = réponse.headers?.get?.('x-vercel-mitigated') || '';
    const suffixe = mitigation ? ` — défi de sécurité du CDN Vercel (X-Vercel-Mitigated: ${mitigation})` : '';
    return { html: null, erreur: `HTTP ${réponse.status}${suffixe}` };
  }

  const corps = await réponse.text();
  if (!/<html[\s>]/i.test(corps)) {
    return { html: null, erreur: "la réponse n'est pas du HTML (le site ne sert pas sa page d'accueil)" };
  }
  return { html: corps, erreur: null };
}

/**
 * Interroge la production jusqu'à ce qu'elle annonce une révision acceptée.
 *
 * Les tentatives absorbent le déploiement en cours ; elles ne transforment pas un
 * refus en succès, elles attendent que la réponse change.
 *
 * @param {object} options `base`, `attendues`, `fetchImpl`, `tentatives`, `delai`.
 * @returns {Promise<{ok: boolean, journal: string[]}>} Le verdict et chaque essai.
 */
export async function verifierServie({
  base,
  attendues = [],
  fetchImpl = fetch,
  tentatives = 9,
  delai = 20,
  dormir = (ms) => new Promise((r) => setTimeout(r, ms * 1000)),
  log = console.log,
  interrogerImpl = interroger,
} = {}) {
  const journal = [];
  const total = Math.max(1, Number(tentatives) || 1);
  for (let tentative = 1; tentative <= total; tentative += 1) {
    const { html, erreur } = await interrogerImpl(base, { fetchImpl });
    const verdict = html === null
      ? {
        ok: false,
        message: `aucune réponse exploitable de ${base} : ${erreur} — rien ne peut être affirmé sur le déploiement.`,
      }
      : comparer(revisionServie(html), attendues);
    journal.push(`tentative ${tentative}/${total} : ${verdict.message}`);
    log(`  ${journal[journal.length - 1]}`);
    if (verdict.ok) return { ok: true, journal };
    if (tentative < total) await dormir(delai);
  }
  return { ok: false, journal };
}

/**
 * @param {object} [options]
 * @param {string} [options.base] Site interrogé (défaut `SITE_ORIGIN`).
 * @returns {{ok: boolean, erreur: string|null}} `erreur` : le refus d'invoquer.
 */
export function baseAutorisée(base) {
  const propre = String(base ?? '').trim().replace(/\/+$/, '');
  if (!propre) return { ok: false, erreur: 'aucune base à interroger' };
  if (isLoopbackUrl(propre)) {
    return {
      ok: false,
      erreur:
        `base locale (${propre}) : ce garde mesure la PRODUCTION (ce que Vercel sert) ; ` +
        "un build de poste ne dit rien de la révision déployée.",
    };
  }
  return { ok: true, erreur: null };
}

/**
 * Lit les options de la ligne de commande, sans jamais conclure sur du vide.
 *
 * @param {string[]} argv Arguments (sans `node`/le script).
 * @returns {{base: string, attendues: string[], tentatives: number, delai: number}}
 */
export function lireOptions(argv = []) {
  const valeurDe = (nom) => {
    const index = argv.indexOf(nom);
    return index === -1 ? '' : String(argv[index + 1] ?? '');
  };
  const nombreDe = (nom, défaut) => {
    const brut = valeurDe(nom);
    return brut ? Number(brut) : défaut;
  };
  const { valides, invalides } = revisionsAttendues([valeurDe('--attendu'), valeurDe('--sinon')]);
  if (invalides.length) {
    throw new Error(
      `--attendu/--sinon ne sont pas des identifiants de commit (${invalides
        .map((v) => JSON.stringify(v))
        .join(', ')}) : ce garde compare la révision servie à un commit, il ne peut pas ` +
        'conclure sans lui.'
    );
  }
  if (valides.length === 0) {
    throw new Error(
      '--attendu manquant : passer le commit poussé (et `--sinon` le dernier commit ayant ' +
        'touché frontend/) — sans quoi rien ne dit à quoi comparer la production.'
    );
  }
  return {
    base: valeurDe('--base') || SITE_ORIGIN,
    attendues: valides,
    tentatives: nombreDe('--tentatives', 9),
    delai: nombreDe('--delai', 20),
  };
}

export async function main(argv = process.argv.slice(2)) {
  let options;
  try {
    options = lireOptions(argv);
  } catch (error) {
    console.error(`::error title=Frontend servi — invocation impossible::${error.message}`);
    return 2;
  }

  const base = baseAutorisée(options.base);
  if (!base.ok) {
    console.error(`::error title=Frontend servi — invocation impossible::${base.erreur}`);
    return 2;
  }

  console.log(`Révisions acceptées : ${options.attendues.map((r) => r.slice(0, 12)).join(' ou ')}`);
  console.log(`Interrogation de ${options.base}/…`);
  const { ok, journal } = await verifierServie(options);
  if (ok) {
    console.log(
      `::notice title=Frontend servi::${journal[journal.length - 1]} — le site sert bien un ` +
        'artefact construit depuis main.'
    );
    return 0;
  }

  const dernier = journal[journal.length - 1] || 'aucune tentative';
  if (dernier.includes('AUCUNE révision')) {
    console.log(`::error title=Révision du frontend non annoncée::${dernier}`);
  } else if (dernier.includes('aucune réponse exploitable')) {
    console.log(`::error title=Frontend injoignable::${dernier}`);
  } else {
    console.log(`::error title=Révision du frontend ≠ attendue::${dernier}`);
  }
  return 1;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  process.exitCode = await main();
}

File: backend/scripts/check_deployed_revision.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Le backend SERVI est-il bien celui de ce commit ?

── Le trou que ce garde ferme ──────────────────────────────────────────────
Le job `deploy-fly` filtrait sur `backend/**` : un push qui ne touchait que le
frontend laissait ses étapes de déploiement en `skipped` et le job passait au
vert. « main est vert » pouvait donc signifier deux choses opposées — « le
backend déployé est celui de ce commit » et « rien n'a été déployé, le service
tourne la version précédente » — sans qu'aucun job ne puisse les distinguer. Le
20/09/2026 le job s'est mis à ÉCRIRE laquelle des deux, ce qui rend le ✓ lisible
mais ne VÉRIFIE rien : une déclaration n'est pas une preuve.

Le service répond désormais à la question qu'on lui posait à sa place :
`/health` publie `revision` (le commit dont l'image a été construite, injecté au
build par `--build-arg KOJO_GIT_SHA`). Ce garde interroge la production et
compare cette réponse à la révision ATTENDUE — celle que ce push aurait dû
déployer :

  • push qui touche `backend/**` (ou le workflow) → l'attendu est le commit poussé ;
  • push qui n'y touche pas → l'attendu est le dernier commit qui y a touché
    (déploiement sauté : le service DOIT tourner cette révision-là, et le job
    échoue si ce n'est pas le cas).

── Ce que ce garde refuse, et pourquoi il le NOMME ─────────────────────────
Trois pannes distinctes, trois corrections distinctes — un seul « échec » les
confondrait :
  1. `revision` vaut « inconnue » : l'image servie est antérieure à ce mécanisme,
     ou elle a été construite sans `--build-arg`. Correction : redéployer.
  2. `revision` est un autre commit : le déploiement n'a pas abouti, ou une
     machine ancienne répond encore. Correction : regarder le job, redéployer.
  3. Aucune réponse exploitable : le service ne répond pas (ou pas du JSON), donc
     RIEN ne peut être affirmé sur le déploiement. Correction : regarder Fly.

Un garde qui ne peut pas savoir ce qu'il compare ÉCHOUE (code 2) au lieu de
passer : `--attendu` vide, ou une valeur trop courte pour être un identifiant de
commit, sont des erreurs d'invocation, pas des succès.

Usage :
    python3 backend/scripts/check_deployed_revision.py --attendu <sha>
Optionnel :
    --url https://api.kojoforafrica.cc.cd/health   (défaut : env KOJO_BACKEND_HEALTH_URL)
    --tentatives 6 --delai 20                      (attente d'un déploiement en cours)
"""
from __future__ import annotations

import argparse
import json
import os
import sys
import time
import urllib.error
import urllib.request

# Adresse par défaut : la même que fly.toml (BACKEND_PUBLIC_URL), jamais un
# hôte interne — ce garde mesure ce qu'un client reçoit.
URL_PAR_DEFAUT = os.environ.get(
    "KOJO_BACKEND_HEALTH_URL", "https://api.kojoforafrica.cc.cd/health"
).strip()

# Longueur minimale d'un SHA de commit : en dessous, ce n'est pas un identifiant
# de commit mais une chaîne quelconque — comparer là-dessus serait un faux vert.
LONGUEUR_SHA_MIN = 7

# Valeurs qui ne sont pas une révision : le défaut du Dockerfile, et le vide.
REVISIONS_ABSENTES = {"", "inconnue", "unknown", "none", "null"}

# Un refus qui plante n'est pas un refus qui NOMME. Sous Windows la console est
# en cp1252 : les accents et «≠» y lèvent UnicodeEncodeError, donc le message
# (servi vs attendu) mourait dans une traceback au lieu d'être lu — la CI, en
# UTF-8 partout, ne le voit jamais (même classe de défaut que test_cors_origins,
# cf. AGENTS.md). On garde l'encodage de la console et on remplace seulement ce
# qu'elle ne sait pas écrire : le verdict et les deux SHA, eux, sont ASCII.
for _flux in (sys.stdout, sys.stderr):
    try:
        _flux.reconfigure(errors="replace")
    except (AttributeError, ValueError):  # flux redirigé/remplacé : rien à faire
        pass


def normaliser(valeur) -> str:
    """La valeur est-elle une RÉVISION ? Vide et « inconnue » ne le sont pas.

    Une seule fonction pour les deux côtés de la comparaison : sans ça, un
    « inconnue » attendu (ou servi) pourrait être comparé à un « inconnue » reçu
    et se déclarer concordant — le faux vert exact que ce garde existe pour
    fermer.
    """
    if not isinstance(valeur, str):
        return ""
    valeur = valeur.strip()
    return "" if valeur.lower() in REVISIONS_ABSENTES else valeur


def revision_annoncee(payload) -> str:
    """La révision publiée par le service, ou '' si la réponse n'en porte pas."""
    if not isinstance(payload, dict):
        return ""
    return normaliser(payload.get("revision"))


def comparer(attendue: str, servie: str) -> tuple[bool, str]:
    """Le verdict, avec le coupable nommé dans chaque cas."""
    attendue = normaliser(attendue)
    servie = normaliser(servie)

    if not servie:
        return False, (
            "le backend servi n'annonce AUCUNE révision (champ `revision` absent ou "
            "« inconnue ») : l'image déployée est antérieure à ce mécanisme, ou elle a "
            "été construite sans `--build-arg KOJO_GIT_SHA`. Redéployer le backend "
            "(job deploy-fly) — tant que ce champ se tait, rien ne prouve quelle "
            "version tourne."
        )
    if servie == attendue:
        return True, "révision servie = attendue (%s)" % servie[:12]
    return False, (
        "révision servie ≠ attendue : servi=%s attendu=%s — le backend déployé n'est "
        "pas celui de ce commit (déploiement non abouti, ou machine ancienne encore "
        "répondante)." % (servie[:12], attendue[:12] or "(aucune)")
    )


def interroger(url: str, fetch=None, timeout: float = 20.0):
    """(payload, erreur) — l'erreur est un message ou None. Aucune exception ne sort."""
    fetch = fetch or _fetch_defaut
    try:
        statut, corps = fetch(url, timeout)
    except urllib.error.HTTPError as exc:  # pragma: no cover - dépend du réseau
        return None, "HTTP %s" % exc.code
    except Exception as exc:  # réseau, DNS, TLS, timeout…
        return None, "%s: %s" % (type(exc).__name__, exc)
    if statut != 200:
        return None, "HTTP %s (attendu 200)" % statut
    try:
        return json.loads(corps), None
    except ValueError:
        return None, "réponse non JSON (le service ne répond pas /health)"


def _fetch_defaut(url: str, timeout: float):
    with urllib.request.urlopen(url, timeout=timeout) as reponse:
        return reponse.status, reponse.read()


def verifier_servi(
    url: str,
    attendue: str,
    fetch=None,
    tentatives: int = 6,
    delai: float = 20.0,
    dormir=time.sleep,
    log=print,
) -> tuple[bool, list[str]]:
    """Interroge le service jusqu'à ce qu'il annonce la révision attendue.

    Les tentatives absorbent le déploiement en cours (une image qui vient d'être
    poussée met quelques secondes à répondre) ; elles ne transforment pas un refus
    en succès, elles attendent que la réponse change.
    """
    journal: list[str] = []
    verdict, message = False, "aucune tentative"
    for tentative in range(1, max(1, tentatives) + 1):
        payload, erreur = interroger(url, fetch=fetch)
        if payload is None:
            verdict, message = False, (
                "aucune réponse exploitable de %s : %s — rien ne peut être affirmé sur "
                "le déploiement." % (url, erreur)
            )
        else:
            verdict, message = comparer(attendue, revision_annoncee(payload))
        journal.append("tentative %d/%d : %s" % (tentative, tentatives, message))
        log("  %s" % journal[-1])
        if verdict:
            return True, journal
        if tentative < tentatives:
            dormir(delai)
    return False, journal


def _valider_attendu(attendu: str) -> str:
    """Refuse une invocation qui ne dit pas quoi comparer (code 2, pas un succès)."""
    brut = (attendu or "").strip()
    attendu = normaliser(brut)
    if len(attendu) < LONGUEUR_SHA_MIN:
        print(
            "::error::--attendu manquant, générique ou trop court (%r) : ce garde compare "
            "la révision servie à un identifiant de commit, il ne peut pas conclure sans "
            "lui. Passer le SHA attendu (commit poussé, ou dernier commit qui a touché "
            "backend/**)." % brut,
            file=sys.stderr,
        )
        raise SystemExit(2)
    return attendu


def main(argv=None) -> int:
    analyseur = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    analyseur.add_argument("--attendu", default=os.environ.get("KOJO_ATTENDU", ""),
                           help="révision attendue (SHA du commit)")
    analyseur.add_argument("--url", default=URL_PAR_DEFAUT,
                           help="URL du health check à interroger")
    analyseur.add_argument("--tentatives", type=int, default=6)
    analyseur.add_argument("--delai", type=float, default=20.0)
    options = analyseur.parse_args(argv)

    attendu = _valider_attendu(options.attendu)
    print("Révision attendue : %s" % attendu)
    print("Interrogation de %s…" % options.url)

    ok, journal = verifier_servi(
        options.url, attendu,
        tentatives=options.tentatives, delai=options.delai,
    )
    if ok:
        print("::notice title=Backend servi::%s — le service tourne bien le commit attendu."
              % journal[-1])
        return 0
    erreur = journal[-1]
    if "aucune réponse exploitable" in erreur:
        print("::error title=Backend injoignable::%s (URL %s)" % (erreur, options.url))
    elif "AUCUNE révision" in erreur:
        print("::error title=Révision non annoncée::%s" % erreur)
    else:
        print("::error title=Révision déployée ≠ attendue::%s" % erreur)
    return 1


if __name__ == "__main__":
    sys.exit(main())
